lib/Optimizer/Transforms/PruneCtrlRelations.cpp

```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_PRUNECTRLRELATIONS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "prune-ctrl-relations"
+
+using namespace mlir;
+
+namespace {
+/// Pattern to convert any gate operator, which may have control operands that
+/// are wires, to the same gate operator with `!quake.control` type operands.
+template <typename OP>
+class MakeControl : public OpRewritePattern<OP> {
+public:
+  using Base = OpRewritePattern<OP>;
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(OP op,
+                                PatternRewriter &rewriter) const override {
+    auto *ctx = rewriter.getContext();
+    if (op.getControls().empty())
+      return failure();
+    auto wireTy = quake::WireType::get(ctx);
+    bool nothingToDo = [&]() {
+      for (auto cv : op.getControls())
+        if (cv.getType() == wireTy)
+          return false;
+      return true;
+    }();
+    if (nothingToDo)
+      return failure();
+
+    // At this point `op` has (some) wires in control positions. Let's convert
+    // this op to remove the artificial constraints.
+    auto coarity = op.getWires().size();
+    auto loc = op.getLoc();
+    auto ctrlTy = quake::ControlType::get(ctx);
+    SmallVector<Value> newCtrls;
+
+    // For each wire control, convert the wire to a control with a ToControlOp.
+    // this also reduces the coarity of the new op.
+    for (auto cv : op.getControls()) {
+      if (cv.getType() == wireTy) {
+        Value input;
+        if (auto fromCtrl = cv.template getDefiningOp<quake::FromControlOp>()) {
+          input = fromCtrl.getCtrlbit();
+        } else {
+          input = rewriter.template create<quake::ToControlOp>(loc, ctrlTy, cv);
+        }
+        newCtrls.push_back(input);
+        coarity--;
+      } else {
+        newCtrls.push_back(cv);
+      }
+    }
+
+    // Create a copy of `op` with the correct coarity and with the control wires
+    // each now passing through a ToControlOp.
+    SmallVector<Type> wireTys{coarity, wireTy};
```

Comment:
Good question. So to have reached this point, there is at least 1 control of type WireType. See line 42. An invariant that is verified by the IR verifier is that wires in == wires out. Line 50 sets `coarity` to the total number of wires out, which is the same as the number in. Because of line 44, `coarity >= 1` must be true. The decrement on line 66 will be for all the wires in the controls, `c`, where `c <= coarity` must be true. `coarity` is the wires in controls and targets. Since `c <= coarity`, `coarity - c >= 0`.

Reply:
> An invariant that is verified by the IR verifier is that wires in == wires out.

I think this is the part that I was missing.

> All that said, I could add an assertion here at line 66 to make sure coarity is not 0 before decrementing.

I'm ok either way...I'll leave it up to you.